### PR TITLE
MG-11 - Implementation of messaging feature on the UI

### LINKED ISSUE #11: Implement Send Message and Read Message feature relocation
### Is there an existing feature request for this?

- [X] I have searched the existing feature requests

### Is your feature request related to a problem? Please describe.

_No response_

### Describe the feature you are requesting, as well as the possible use case(s) for it.

Figure out the ideal location to implement the send message feature on the UI and implement it. Also, relocate the read message feature to be in the same location as the send message feature to make it easier to move to that page without having to copy the channel ID and thing Key. 

### Indicate the importance of this feature to you.

- [X] Must-have
- [ ] Should-have
- [ ] Nice-to-have

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by ianmuchyri**
# What type of PR is this?

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Dependency Update

## What does this do?
Proper implementation of the messaging feature (both send and read messages from db) in the UI. 
Previously, we did not have the send message feature, and for the read message a user had to manually enter the channel ID and thing Key to be able to read messages stored in the database.
In this implementation, the messaging features are combined into one page, which is accessed via the thingsChannel page. In this page, the user clicks on the messaging icon and they are moved to the messages page. Here they can be able to view all the messages stored in the database that have been sent to that channel.They can also be able to send a test message to the channel using the device they had accessed the channel with.

## Which issue(s) does this PR fix/relate to?

- Resolves #11 

## Have you included tests for your changes?

- [ ] Yes
- [x] No, and this is why: 

## Did you document any new/modified functionality?

- [ ] Yes
- [x] No, and this is why: 

### Notes
image
image
image
image
image

### CODE REVIEW

**Review by rodneyosodo — CHANGES_REQUESTED** [review]
Pagination doesn't work. Publish 100 messages and try and go to the second page

File: ui/service.go [hunk 1]
```diff
@@ -1572,66 +1577,62 @@ func (us *uiService) ListUserGroupChannels(token, groupID string, page, limit ui
 	return btpl.Bytes(), nil
 }
 
-func (gs *uiService) Publish(token, thKey string, msg *messaging.Message) error {
-	if err := gs.sdk.SendMessage(msg.Channel, string(msg.Payload), thKey); err != nil {
-		return errors.Wrap(err, ErrFailedPublish)
+func (us *uiService) Messages(token, chID, thKey string, page, limit uint64) ([]byte, error) {
+	var thingKey string
```
Comment (rodneyosodo) [inline ui/service.go:1581]: No need to read messages using thing. You can use the user token to read messages from a channel https://github.com/absmach/magistrala-old/pull/199

File: ui/api/requests.go [hunk 2]
```diff
@@ -623,40 +622,54 @@ func (req updateGroupStatusReq) validate() error {
...
 	}
-	if req.Msg.Payload == nil {
+	if req.Message == "" {
 		return errMissingPayload
 	}
 	return nil
 }
 
-type readMessageReq struct {
+type messagesReq struct {
```
Comment (rodneyosodo) [inline ui/api/requests.go:648]: revert its name to `readMessageReq`

File: ui/service.go [hunk 3]
```diff
@@ -274,9 +273,10 @@ type Service interface {
 	ListUserGroupChannels(token, groupID string, page, limit uint64) (b []byte, err error)
 
 	// Publish facilitates a thing publishin messages to a channel.
-	Publish(token, thKey string, msg *messaging.Message) error
-	// ReadMessage facilitates a thing reading messages published in a channel.
-	ReadMessage(token, chID, thKey string, page, limit uint64) ([]byte, error)
+	Publish(token, chID, thKey, message string) error
+	// Messages retrieves messages published in a channel.
+	Messages(token, chID, thKey string, page, limit uint64) ([]byte, error)
```
Comment (rodneyosodo) [inline ui/service.go:278]: Revert its name to `ReadMessages` same as other occurrences

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

**Comment by ianmuchyri** [comment]
> Pagination doesn't work. Publish 100 messages and try and go to the second page

will be fixed in this pr: https://github.com/absmach/magistrala-ui/pull/77

**Comment by ianmuchyri** [comment]
> Pagination doesn't work. Publish 100 messages and try and go to the second page

we don't have pagination in Magistrala `readMessage`

**Comment by ianmuchyri** [comment]
> > Pagination doesn't work. Publish 100 messages and try and go to the second page
> 
> we don't have pagination in Magistrala `readMessage`

It will be resolved by this pr: https://github.com/absmach/magistrala/pull/2055

File: ui/service.go [hunk 4]
```diff
@@ -281,9 +280,10 @@ type Service interface {
 	ListUserGroupChannels(token, groupID string, page, limit uint64) (b []byte, err error)
 
 	// Publish facilitates a thing publishin messages to a channel.
-	Publish(token, thKey string, msg *messaging.Message) error
-	// ReadMessage facilitates a thing reading messages published in a channel.
-	ReadMessage(token, chID, thKey string, page, limit uint64) ([]byte, error)
+	Publish(token, chID, thKey, message string) error
+	// ReadMessages retrieves messages published in a channel.
+	ReadMessages(token, chID, thKey string, page, limit uint64) ([]byte, error)
```
Comment (rodneyosodo) [inline ui/service.go:285]:
```suggestion
	ReadMessages(token, chID string, page, limit uint64) ([]byte, error)
```
No need for thingKey

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: ui/service.go (hunk 4 above)
Comment (ianmuchyri) [reply ui/service.go:285]: the thing key is passed back to the HTML template to be used by the t`est channel modal` for publishing to the channel

**Comment by ianmuchyri** [comment]
the CI failure will be resolved by this https://github.com/absmach/agent/pull/4

File: ui/api/transport.go [hunk 5]
```diff
@@ -61,6 +60,7 @@ const (
 	groupsItem              = "groups"
 	accessTokenKey          = "token"
 	refreshTokenKey         = "refresh_token"
+	baseName                = "test-"
```
Comment (rodneyosodo) [inline ui/api/transport.go:63]: No need for this as it prefix messages from UI

File: ui/api/transport.go [hunk 6]
```diff
@@ -1608,45 +1608,56 @@ func decodeGroupStatusUpdate(_ context.Context, r *http.Request) (interface{}, e
...
-		Payload:  []byte(r.Form.Get("message")),
-		Created:  time.Now().UnixNano(),
+	currentTime := time.Now().Unix()
+	floatValue, err := strconv.ParseFloat(r.PostFormValue("value"), 64)
+	if err != nil {
+		return nil, err
 	}
 
-	token, err := tokenFromCookie(r, "token")
+	message := struct {
```
Comment (rodneyosodo) [inline ui/api/transport.go:1622]: Add this on `publishReq` struct

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

**Review by rodneyosodo — APPROVED** [review]

**Review by dborovcanin — CHANGES_REQUESTED** [review]
Please resolve conflicts so we can test and review.

**Pull request merged**